sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/RecoverableReactorConnection.java

```diff
@@ -0,0 +1,249 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.amqp.implementation;
+
+import com.azure.core.amqp.AmqpRetryOptions;
+import com.azure.core.amqp.AmqpRetryPolicy;
+import com.azure.core.amqp.AmqpShutdownSignal;
+import com.azure.core.amqp.exception.AmqpErrorContext;
+import com.azure.core.amqp.exception.AmqpException;
+import com.azure.core.util.logging.ClientLogger;
+import reactor.core.Disposable;
+import reactor.core.publisher.Mono;
+import reactor.util.retry.Retry;
+
+import java.time.Duration;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Supplier;
+
+import static com.azure.core.amqp.implementation.ClientConstants.CONNECTION_ID_KEY;
+import static com.azure.core.amqp.implementation.ClientConstants.INTERVAL_KEY;
+
+public final class RecoverableReactorConnection<C extends ReactorConnection> implements Disposable {
```

Comment:
I am thinking about the naming for this class. It is more like a connection container/manager here. Do you think we need to add a suffix like `Container` or `Manager`? 

Or we can keep using `RecoverableReactorConnection`, but let it extends the `ReactorConnection`. We still maintain a delegate currentConnection and wrap it with cacheInvalidateIf operator and retry policy. We involve the relationship for all Connection classes, so that we can treat `RecoverableReactorConnection` as a connection as well. But this would require more refactoring work and may break some internal API. Ideally, the class hierarchy could be: 
```mermaid
classDiagram
ReactorConnection <|-- RecoverableReactorConnection
RecoverableReactorConnection <|-- EventHubReactorAmqpConnection
RecoverableReactorConnection <|-- ServiceBusReactorAmqpConnection
RecoverableReactorConnection: +currentConnection
RecoverableReactorConnection: +get()
EventHubReactorAmqpConnection: +currentConnection
EventHubReactorAmqpConnection: +get()
ServiceBusReactorAmqpConnection: +currentConnection
ServiceBusReactorAmqpConnection: +get()
```
This is just an idea I got from the naming. What do you think?

Reply:
Hi Kun @liukun-msft, I agree with you; the "RecoverableReactorConnection" is nothing but a _container_ acting as a "single element cache", where the responsibilities are - 

 1. _Caches_ a single connection.
 2. Refresh the cache when the _cached_ element (connection) is expired.
 3. Clean up / dispose of the _cache_ when it is no longer needed.

All three responsibilities are about managing a "single element (connection) cache" i.e. a common theme is _cache_, so a possible name for this internal _container_ type that probably makes it readable for future devs is "ReactorConnectionCache", with the current contract as-

| API | Description |
|:------|:--------|
| ReactorConnectionCache::Ctr(ConnectionSupplier&lt;ReactorConnection&gt;..)  | Ctr API taking supplier that supplies connection when the cache needs to be refreshed |
| ReactorConnectionCache::get() | API to read from the cache (The read can indirectly refresh the cache). |
| ReactorConnectionCache::dispose() | API to Clean up / dispose of the cache. |

What do you think?

[by the way, the diagrams are nice! I feel that the second apporach of viewing "RecoverableReactorConnection" as a connection (for delegation) rather than a container (the first apporach) could be overwhelming]

Reply:
Just for future reference, I thought about the prefix _Pool_ . Pool usually means a pool of multiple elements (not single), e.g., HTTP-connections, that are reusable, but only one consumer can own an element at a time, and it needs to be released for another consumer to obtain. The common connection-sharing usecase in messaging differs from that common understanding of Pool.

Reply:
Thank you Anu, I agree with naming it "ReactorConnectionCache", it's more accurate and readable!

Reply:
Renamed to "ReactorConnectionCache".